File: src/options.py
from dataclasses import dataclass
from enum import Enum

import numpy as np


class OptionType(Enum):
    CALL = "call"
    PUT = "put"


class PositionType(Enum):
    BUY = "buy"
    SELL = "sell"


@dataclass
class OptionPosition:
    type: OptionType  # OptionType Enum
    position: PositionType  # PositionType Enum

    quantity: float
    strike_price: float  # Strike price
    premium_per_item: float  # Premium paid (if BUY) or received (if SELL)

    @property
    def label(self):
        return f"{self.type.value} {self.position.value} {self.quantity} @ {self.strike_price}"

    @property
    def total_premium(self):
        return self.quantity * self.premium_per_item

    def payoff(self, price: float | np.ndarray):
        """
        Calculate the payoff for a single option leg.
        S: array of underlying prices at expiration
        """
        if self.type == OptionType.CALL:
            payoff = np.maximum(price - self.strike_price, 0) * self.quantity
        elif self.type == OptionType.PUT:
            payoff = np.maximum(self.strike_price - price, 0) * self.quantity
        else:
            raise ValueError("Invalid option type")

        if self.position == PositionType.BUY:
            return payoff - self.total_premium
        elif self.position == PositionType.SELL:
            return -payoff + self.total_premium
        else:
            raise ValueError("Invalid position type")

File: src/options.ts
// Equivalent to Python's enum
export enum OptionType {
  CALL = "Call",
  PUT = "Put",
}

export enum PositionType {
  BUY = "Buy",
  SELL = "Sell",
}

// Equivalent to Python's dataclass
export class OptionPosition {
  type: OptionType;
  position: PositionType;
  quantity: number;
  strike_price: number;
  premium_per_item: number;

  constructor(
    type: OptionType,
    position: PositionType,
    quantity: number,
    strike_price: number,
    premium_per_item: number,
  ) {
    this.type = type;
    this.position = position;
    this.quantity = quantity;
    this.strike_price = strike_price;
    this.premium_per_item = premium_per_item;
  }

  // Equivalent to Python's property
  get label(): string {
    return `${this.type} ${this.position} ${this.quantity} @ ${this.strike_price}. $${this.total_premium}`;
  }

  // Equivalent to Python's property
  get total_premium(): number {
    return this.quantity * this.premium_per_item;
  }

  // Equivalent to Python's method with type annotation
  payoff(price: number | number[]): number | number[] {
    if (Array.isArray(price)) {
      return price.map((p) => this.payoff(p) as number);
    }

    let payoff: number;
    if (this.type === OptionType.CALL) {
      payoff = Math.max(price - this.strike_price, 0) * this.quantity;
    } else if (this.type === OptionType.PUT) {
      payoff = Math.max(this.strike_price - price, 0) * this.quantity;
    } else {
      throw new Error("Invalid option type");
    }

    if (this.position === PositionType.BUY) {
      return payoff - this.total_premium;
    } else if (this.position === PositionType.SELL) {
      return -payoff + this.total_premium;
    } else {
      throw new Error("Invalid position type");
    }
  }
}
